### ISSUE TITLE
Cannot override values when spreading

### ISSUE THREAD

**Issue by villesau** [facebook/flow#6108]
[Try flow link](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCCYAvGAN6phgCGAXGAHYCuAtgEZ4BOqAvgNzps+MACFSFKrQYBnDFwCWjAOZ9BqAMZxGcsBjxyAYowkAKJTmYYGhAJQNxJAHxhT5AHSeLVgDTSwAOTMABZgcKF0wAG8tvxAA)

```
type A = {
  a: number
};

type B = {
  a: string
};

const testFn = (input: A): B => ({...input, a: 'uh oh :/'});
```

As you can see, the property `a` is overridden with string in the function. Both, input and output are type checked, but flow complains that output and input are not compatible. This is very basic and idiomatic ES6 behaviour and should be fully supported.

**Comment by vicapow** [comment]
You can override values when spreading it's just that when the two types differ, flow infers the type to be an union of the two. This seems like reasonable behavior to me.

````js
const a: { a: (number | string) } = {...{a: 10}, a: 'no erro/'};
```

**Comment by villesau** [comment]
I think it's clear from the code that the result type of `a` can't be number anymore and flow should be able to infer it as string. It just can't be number after spreading anymore.

**Comment by theres** [comment]
Imho this should work. If you need to replace a few of fields out multiple, then you can use rest operator as a _good engough_  [workaround](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCCYAvGAN6phgCGAXGAHYCuAtgEZ4BOANFWOwYBnDFwCWjAOZ9qAYwbs4cGHhqNUAXwDc6bPjAAhUhX70wI8VJkDhoidP7yBSlWs07UsuIxFgMeEQAxRmMACnIaHjAAOliuAIwNBkIASgYjEgA+MHDY6PiRKLMAcgQ4LgBrIQBCYo0UrSA):

```js
type A = {
  a: number,
  b: string,
  c: boolean
};

type B = {
  a: string,
  b: string,
  c: boolean
};

const testFn = ({a, ...rest}: A): B => ({...rest, a: 'works!'});
```

**Comment by wchargin** [comment]
@mrkev imho, this is not a feature request: it’s a bug. Flow’s understanding of the semantics of spread is simply incorrect, and inconsistent with its (correct) understanding of how field overriding works _without_ spread: namely, `{a: 1, a: "won"}` correctly has type `{a: string}` and not `{a: number}`.

**Comment by mrkev** [comment]
Yup, looks like it's buggy

**Comment by villesau** [comment]
Managed to make a PR to fix this: https://github.com/facebook/flow/pull/7298

**Comment by nmote** [comment]
This has been fixed and will typecheck in Flow v0.111.0
